nameko/events.py

```diff
@@ -183,10 +168,63 @@ def __init__(self, source_service, event_type, handler_type=SERVICE_POOL,
         super(EventHandler, self).__init__(
             queue=None, requeue_on_error=requeue_on_error)
 
+    @property
+    def broadcast_identifier(self):
+        """ A unique string to identify a service instance for `BROADCAST`
+        type handlers.
+
+        The `broadcast_identifier` is appended to the queue name when the
+        `BROADCAST` handler type is used. It must uniquely identify service
+        instances that receive broadcasts.
+
+        The default `broadcast_identifier` is a uuid that is set when the
+        service starts. It will change when the service restarts, meaning
+        that any unconsumed messages that were broadcast to the 'old' service
+        instance will not be received by the 'new' one. ::
+
+            @property
+            def broadcast_identifier(self):
+                # use a uuid as the identifier.
+                # the identifier will change when the service restarts and
+                # any unconsumed messages will be lost
+                return uuid.uuid4().hex
+
+        The default behaviour is therefore incompatible with reliable delivery.
+
+        An alternative `broadcast_identifier` that would survive service
+        restarts is ::
+
+            @property
+            def broadcast_identifier(self):
+                # use the machine hostname as the identifier.
+                # this assumes that only one instance of a service runs on
+                # any given machine
+                return socket.gethostname()
+
+        If neither of these approaches are appropriate, you could read the
+        value out of a configuration file ::
+
+            @property
+            def broadcast_identifier(self):
+                return self.config['SERVICE_IDENTIFIER']  # or similar
+
+        Broadcast queues are exclusive to ensure that `broadcast_identifier`
+        values are unique.
+        """
+        if self.reliable_delivery and self.handler_type is BROADCAST:
```

Comment:
i think the strangeness is caused by a problem "higher up", namely trying to have a single class for both "service pool" and "broadcast" events (ignoring singleton for now)

assuming any change to that is out of scope for now

the initial implementation had a kwarg that only made sense for certain values of handler_type. having a method is less bad, but we still have this method that only make sense if the class was initialised with a certain parameter value

we are extending `EventHandler` and added a protocol which means it calls `broadcast_handler` if it's in broadcast mode. i think calling that method "out of turn" is fine to be "undefined behaviour". returning `None` as you suggest is fine

however, i think my (new) objection to the property decorator is that because we use `getmembers`, the method (or a new implementation in user code) gets called at unexpected times. so regardless of whether we can get this particular implementation of `broadcast_handler` to handle that, i'd prefer to stay away from descriptors until we can make our inspection not invoke them